rllib/core/optim/tests/test_rl_optimizer.py

```diff
@@ -0,0 +1,378 @@
+        self._rl_optimizer = BCTorchOptimizer(self._module, optimizer_config)
+
+    @staticmethod
+    def on_after_compute_gradients(
+        gradients_dict: Mapping[str, Any]
+    ) -> Mapping[str, Any]:
+        """Called after gradients have been computed.
+
+        Args:
+            gradients_dict (Mapping[str, Any]): A dictionary of gradients.
+
+        Note the relative order of operations looks like this:
+            fwd_out = forward_train(batch)
+            loss = compute_loss(batch, fwd_out)
+            gradients = compute_gradients(loss)
+            ---> post_processed_gradients = on_after_compute_gradients(gradients)
+            apply_gradients(post_processed_gradients)
+
+        Returns:
+            Mapping[str, Any]: A dictionary of gradients.
+        """
+        return gradients_dict
+
+    def compile_results(
+        self,
+        batch: NestedDict,
+        fwd_out: Mapping[str, Any],
+        postprocessed_loss: Mapping[str, Any],
+        post_processed_gradients: Mapping[str, Any],
+    ) -> Mapping[str, Any]:
+        """Compile results from the update.
+
+        Args:
+            batch: The batch that was used for the update.
+            fwd_out: The output of the forward train pass.
+            postprocessed_loss: The loss after postprocessing.
+            post_processed_gradients: The gradients after postprocessing.
+
+        Returns:
+            A dictionary of results.
+        """
+
+        loss_numpy = convert_to_numpy(postprocessed_loss)
+        results = {
+            "avg_reward": batch["rewards"].mean(),
+        }
+        results.update(loss_numpy)
+        results["module_norm"] = model_norm(self._module)
+        return results
+
+    def update(self, batch: NestedDict) -> Mapping[str, Any]:
+        """Perform an update on this Trainer.
+
+        Args:
+            batch: A batch of data.
+
+        Returns:
+            A dictionary of results.
+        """
+        torch_batch = convert_to_torch_tensor(batch)
+        fwd_out = self._module.forward_train(torch_batch)
+        loss = self._rl_optimizer.compute_loss(torch_batch, fwd_out)
+        gradients = self.compute_gradients(loss)
+        post_processed_gradients = self.on_after_compute_gradients(gradients)
+        self.apply_gradients(post_processed_gradients)
+        results = self.compile_results(batch, fwd_out, loss, post_processed_gradients)
+        return results
+
+    def compute_gradients(
+        self, optimization_vars: Mapping[str, Any]
+    ) -> Mapping[str, Any]:
+        """Perform an update on self._module
+
+            For example compute and apply gradients to self._module if
+                necessary.
+
+        Args:
+            optimization_vars: named variables used for optimizing self._module
+
+        Returns:
+            A dictionary of extra information and statistics.
+        """
+        # for torch
+        optimization_vars["total_loss"].backward()
```

Comment:
`total_loss` key here assumes certain pattern on the `_configure_optimizer` output. Let's settle that to something that covers the use-cases discussed above and provide a fix here.